src/epgsql_sock.erl

```diff
@@ -140,25 +149,37 @@ handle_call({standby_status_update, FlushedLSN, AppliedLSN}, _From,
     send(State, ?COPY_DATA, epgsql_wire:encode_standby_status_update(ReceivedLSN, FlushedLSN, AppliedLSN)),
     {reply, ok, State#state{repl_last_flushed_lsn = FlushedLSN, repl_last_applied_lsn = AppliedLSN}};
 
+handle_call({timed, Command, Timeout}, From, State) ->
+  Timer = erlang:start_timer(Timeout, self(), timed_command),
+  #state{queue = Q} = State,
+  Req = {{call, From}, Command},
+  do_command(Command, State#state{queue = queue:in(Req, Q),
+                                  complete_status = undefined,
+                                  timer = Timer});
 handle_call(Command, From, State) ->
-    #state{queue = Q} = State,
-    Req = {{call, From}, Command},
-    command(Command, State#state{queue = queue:in(Req, Q),
-                                 complete_status = undefined}).
+  #state{queue = Q} = State,
+  Req = {{call, From}, Command},
+  do_command(Command, State#state{queue = queue:in(Req, Q),
+                               complete_status = undefined}).
 
 handle_cast({{Method, From, Ref}, Command} = Req, State)
   when ((Method == cast) or (Method == incremental)),
        is_pid(From),
        is_reference(Ref)  ->
     #state{queue = Q} = State,
-    command(Command, State#state{queue = queue:in(Req, Q),
+    do_command(Command, State#state{queue = queue:in(Req, Q),
                                  complete_status = undefined});
 
 handle_cast(stop, State) ->
     {stop, normal, flush_queue(State, {error, closed})};
 
-handle_cast(cancel, State = #state{backend = {Pid, Key},
-                                   sock = TimedOutSock}) ->
+handle_cast(cancel, State) ->
+  do_cancel(State),
+  {noreply, State}.
+
+do_cancel(State) ->
```

Comment:
Please, move `do_cancel` implementation somewhere below `terminate` callback. I think it's not ok to mix internal functions and gen_server's callbacks.